crates/matrix-sdk-crypto/src/olm/group_sessions/share_strategy.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    ops::Deref,
+};
+
+use itertools::{Either, Itertools};
+use ruma::{DeviceId, OwnedDeviceId, OwnedUserId, UserId};
+use serde::{Deserialize, Serialize};
+use tracing::{debug, trace};
+
+use super::OutboundGroupSession;
+use crate::{
+    error::OlmResult, store::Store, types::events::room_key_withheld::WithheldCode,
+    EncryptionSettings, ReadOnlyDevice,
+};
+
+/// Returned by `collect_session_recipients`.
+///
+/// Information indicating whether the session needs to be rotated
+/// (`should_rotate`) and the list of users/devices that should receive
+/// (`devices`) or not the session,  including withheld reason
+/// `withheld_devices`.
+#[derive(Debug)]
+pub(crate) struct CollectRecipientsResult {
+    /// If true the outbound group session should be rotated
+    pub should_rotate: bool,
+    /// The map of user|device that should receive the session
+    pub devices: BTreeMap<OwnedUserId, Vec<ReadOnlyDevice>>,
+    /// The map of user|device that won't receive the key with the withheld
+    /// code.
+    pub withheld_devices: Vec<(ReadOnlyDevice, WithheldCode)>,
+}
+
+struct KeySharingResult {
+    /// The map of user|device that should have access to current room key
+    pub allowed_devices: BTreeMap<OwnedUserId, Vec<ReadOnlyDevice>>,
+    /// The map of user|device that won't receive the room key with the withheld
+    /// code.
+    pub withheld_devices: Vec<(ReadOnlyDevice, WithheldCode)>,
+}
+
+pub(crate) struct CollectRecipientsHelper {}
```

Comment:
Is there any need for this struct? Can't we just have `collect_session_recipients` as a free function?